### PR TITLE
feat: add dynamic slippage to trades

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

Uses the `recommendedSlippage` value provided by the trade quote when it's available, else uses the default value.

Requires https://github.com/shapeshift/lib/pull/1102 to be published.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/lib/issues/1094

## Risk

Small/medium - worst case the `recommendedSlippage` doesn't work and our `limit` is slightly smaller than it should be, causing trades to fail.

## Testing

Trades on swappers that utilise slippage values (currently 0x and THORSwap) should still go through correctly.

### Engineering

The bulk of the engineering review should have already taken place on https://github.com/shapeshift/lib/pull/1102.

### Operations

☝️ 

## Screenshots (if applicable)

N/A

### CODE REVIEW

File: src/components/Trade/hooks/useTradeQuoteService.tsx [hunk 1]
```diff
@@ -179,6 +180,16 @@ export const useTradeQuoteService = () => {
   // Update trade quote
   useEffect(() => setValue('quote', tradeQuote), [tradeQuote, setValue])
 
+  // Set slippage if the quote contains a recommended value, else use the default
+  useEffect(
+    () =>
+      setValue(
+        'slippage',
+        tradeQuote?.recommendedSlippage ? tradeQuote.recommendedSlippage : DEFAULT_SLIPPAGE,
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useTradeQuoteService.tsx:188]: This is where the action happens - if we get a `recommendedSlippage` back on our quote, use it.

**Review by 0xApotheosis — COMMENTED** [review]
Notes for reviewers.

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx [hunk 2]
```diff
@@ -50,7 +50,7 @@ export const ReceiveSummary: FC<ReceiveSummaryProps> = ({
 
   const slippageAsPercentageString = bnOrZero(slippage).times(100).toString()
   const amountAfterSlippage = bnOrZero(amount)
-    .times(1 - slippage)
+    .times(1 - Number(slippage))
```
Comment (gomesalexandre) [inline src/components/Trade/TradeConfirm/ReceiveSummary.tsx:53]: `bn(1).minus(slippage)` might be safer

**Review by gomesalexandre — APPROVED** [review]
image
image

image
image
image

Note: For the last screenshot, I was routed to Mintscan when clicking "View Transaction" and actually had to manually input the Tx in viewblock.io

Not related to this PR, but leaving this here for reference, I somehow managed to got into a state we noticed previously where reacting on Tx change rugs the expected amount and doesn't subtract protocol fees. I didn't manage to reproduce so this may have been a one-off local build issue

image

File: src/components/Trade/TradeConfirm/ReceiveSummary.tsx (hunk 2 above)
Comment (0xApotheosis) [reply src/components/Trade/TradeConfirm/ReceiveSummary.tsx]: I'm all about safety: https://github.com/shapeshift/web/pull/3419/commits/6eac305b76c8cd3cd5ad6cf027309b0c539c1eb2

**Review by gomesalexandre — APPROVED** [review]

**Pull request merged**